src/Rendering/Bs4FormRenderer.php

```diff
@@ -0,0 +1,148 @@
+<?php
+
+/**
+ * This file is part of the Nextras community extensions of Nette Framework
+ *
+ * @license    New BSD License
+ * @link       https://github.com/nextras/forms
+ */
+
+namespace Nextras\Forms\Rendering;
+
+use Nette;
+use Nette\Forms\Controls;
+use Nette\Forms\Form;
+use Nette\Forms\Rendering\DefaultFormRenderer;
+
+/**
+ * FormRenderer for Bootstrap 4 framework.
+ */
+class Bs4FormRenderer extends DefaultFormRenderer
+{
+	/** @var Controls\Button */
+	public $primaryButton = null;
+
+	/** @var bool */
+	private $controlsInit = false;
+
+	/** @var int */
+	private $layout = null;
+
+	public function __construct($layout=FormLayout::HORIZONTAL)
+	{
+		$this->layout = $layout;
+
+		$this->wrappers['controls']['container'] = null;
+		$this->wrappers['pair']['container'] = 'div class="form-group' . ($layout == FormLayout::HORIZONTAL ? ' row' : '') . '"';
+		$this->wrappers['control']['container'] = $layout == FormLayout::HORIZONTAL ? 'div class=col-sm-9' : null;
+		$this->wrappers['label']['container'] = $layout == FormLayout::HORIZONTAL ? 'div class="col-sm-3 col-form-label"' : null;
+		$this->wrappers['control']['description'] = 'small class="form-text text-muted"';
+		$this->wrappers['control']['errorcontainer'] = 'div class=invalid-feedback';
+		$this->wrappers['control']['.error'] = 'is-invalid';
+		$this->wrappers['control']['.file'] = 'form-file';
+		$this->wrappers['error']['container'] = null;
+		$this->wrappers['error']['item'] = 'div class="alert alert-danger" role=alert';
+	}
+
+
+	public function renderBegin()
+	{
+		$this->controlsInit();
+		return parent::renderBegin();
+	}
+
+
+	public function renderEnd()
+	{
+		$this->controlsInit();
+		return parent::renderEnd();
+	}
+
+
+	public function renderBody()
+	{
+		$this->controlsInit();
+		return parent::renderBody();
+	}
+
+
+	public function renderControls($parent)
+	{
+		$this->controlsInit();
+		return parent::renderControls($parent);
+	}
+
+
+	public function renderPair(Nette\Forms\IControl $control)
+	{
+		$this->controlsInit();
+		return parent::renderPair($control);
+	}
+
+
+	public function renderPairMulti(array $controls)
+	{
+		$this->controlsInit();
+		return parent::renderPairMulti($controls);
+	}
+
+
+	public function renderLabel(Nette\Forms\IControl $control)
+	{
+		$this->controlsInit();
+		return parent::renderLabel($control);
+	}
+
+
+	public function renderControl(Nette\Forms\IControl $control)
+	{
+		$this->controlsInit();
+		return parent::renderControl($control);
+	}
+
+
+	private function controlsInit()
+	{
+		if ($this->controlsInit) {
+			return;
+		}
+
+		$this->controlsInit = true;
+
+		if ($this->layout == FormLayout::INLINE) {
+			$this->form->getElementPrototype()->addClass('form-inline');
+		}
+
+		foreach ($this->form->getControls() as $control) {
+			// TODO: remove after https://github.com/nette/forms/pull/209 is available
```

Comment:
PR is merged :)

Reply:
We cannot remove this, until new Nette\Forms is released and we depend on it. I plan to upgrade Nextras\Forms for Nette 3.0 later.